app/models/jellyfish_demo/service/storage.rb

```diff
@@ -0,0 +1,55 @@
+module JellyfishDemo
+  module Service
+    class Storage < ::Service
+      def actions
+        actions = super.merge :terminate
+        actions
+      end
+
+      def provision
+        random_id = 1 + rand(99_999)
+        random_ip = rand(999)
+
+        details = {
+          image_id: product.settings[:image_id],
+          flavor_id: product.settings[:flavor_id],
+          key_name: product.settings[:key_name],
+          subnet_id: product.settings[:subnet_id],
+          instance_id: random_id,
+          public_ip_address: "192.178.0.#{random_ip}"
+        }
+
+        # SAVE PRODUCT DETAILS
+        save_outputs(details, [['image_id', :image_id], ['flavor_id', :flavor_id], ['key_name', :key_name],
```

Comment:
All of the strings match the symbols. You can get a string from a symbol with `to_s`, and get a symbol from a string with `to_sym`.

You're using `defined?` for a value that was declared one statement above. This was copy and pasted from elsewhere?

`if defined? details` isn't doing what you think.

```ruby
defined? 'foobar'  # "expression"
defined? nil       # "nil"
defined? true      # "true"
defined? false     # "false"
defined? TrueClass # "constant"
```